### PR TITLE
Fix clientId issues.

### PR DESCRIPTION
**Pull request by JoshDunnDev**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -540,7 +543,13 @@ api.verify = async ({
...
   if(type === 'nonce') {
     for(const token of tokenValue) {
-      if(token.sha256 === prefixedHash(hash)) {
+      let combinedHash;
+      if(clientId) {
+        combinedHash = prefixedHash(`${hash}:${clientId}`);
+      } else {
+        combinedHash = prefixedHash(hash);
+      }
+      if(token.sha256 === combinedHash) {
```
Comment (dlongley) [inline lib/index.js:552]: So I'm also noticing that this code should have been using `crypto.timingSafeEqual` from line 612 below where the verification is performed. I wonder if we could reorg the code a bit so we don't have to do the combined hash and the comparison twice and so we don't accidentally have bugs like this where we do a constant time compare in one place but not the other.

**Review by dlongley — CHANGES_REQUESTED** [review]
This is good except for the comment ... we should reorganize this code a bit to address it.

File: lib/index.js [hunk 2]
```diff
@@ -225,10 +224,6 @@ api.set = async ({
 
   let tokenTypeValue;
   if(type === 'nonce') {
-    // include old value in the query to ensure that no concurrent
-    // changes occurred via another process
-    const oldValue = tokens.length > 0 ? tokens.slice() : {$in: [null, []]};
-    query['meta.bedrock-authn-token.tokens.nonce'] = oldValue;
```
Comment (dlongley) [inline lib/index.js:231]: Why was this removed? My understanding is that, previously, if another process added a new token or cleared one that was used up that this code would prevent those changes from being dropped. That's probably why `filterExpiredTokens` was also set to `false` ... we may need a different fix here.

File: lib/index.js [hunk 3]
```diff
@@ -536,11 +535,22 @@ api.verify = async ({
...
+  let nonceVerified;
   let tokenValue = record.meta['bedrock-authn-token'].tokens[type];
   const now = new Date();
   if(type === 'nonce') {
     for(const token of tokenValue) {
-      if(token.sha256 === prefixedHash(hash)) {
+      nonceVerified = crypto.timingSafeEqual(
+        new Buffer(token.sha256, 'base64'),
+        new Buffer(combinedHash, 'base64'));
+      if(nonceVerified) {
```
Comment (dlongley) [inline lib/index.js:553]: There's still a bug in this section where -- if the the `combinedHash` doesn't match *any* of the tokens that the `tokenValue` will be set to the array of tokens instead of being undefined, right? It seems like maybe we can return `false` early when that happens? Can you check on that?

File: lib/index.js [hunk 4]
```diff
@@ -593,10 +603,14 @@ api.verify = async ({
...
-    verified = crypto.timingSafeEqual(
-      new Buffer(token.sha256, 'base64'),
-      new Buffer(prefixedHash(hash), 'base64'));
+    if(type === 'nonce') {
+      verified = nonceVerified;
+    } else {
+      // verify hash-based token value
+      verified = crypto.timingSafeEqual(
+        new Buffer(token.sha256, 'base64'),
+        new Buffer(combinedHash, 'base64'));
```
Comment (dlongley) [inline lib/index.js:612]: Can you make a local helper function for this so we can call it in both places?

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 1 above)
Comment (dlongley) [reply lib/index.js]: Yeah, it's pretty good -- if we can create a local helper for the comparison that would finish up the refactor part nicely.

File: lib/index.js (hunk 2 above)
Comment (JoshDunnDev) [reply lib/index.js:231]: Hmm well if we set `filterExpiredTokens` this code does not work. Any suggestions on what to do here then if we need to keep this?

File: lib/index.js (hunk 3 above)
Comment (JoshDunnDev) [reply lib/index.js:539]: I think it might be better to throw an error here rather than return `false` early. What do you think?

Comment (JoshDunnDev) [reply lib/index.js:539]: Well it looks like the test is set up to return `false` if a different hash is passed. So I will keep it that way.

File: lib/index.js (hunk 2 above)
Comment (JoshDunnDev) [reply lib/index.js:231]: `oldValue` is this case is looking at the filtered tokens. I assume we need to get the unfiltered tokens to set is the oldValue so it can query correctly?

Comment (dlongley) [reply lib/index.js:231]: Yes, what this comes down to is that we can't use that `filterExpiredTokens` flag (we can't set it to `true`). We'll have to do the expiration checking ourselves or change something else in the API to enable us to both get the old token value and have it return to use just the non-expired tokens. Perhaps when `filterExpiredTokens` is true we could all of the tokens, the filtered tokens, and the expired tokens in three different properties (as an example). We'll want some kind of elegant solution here that allows us to avoid repeating code.

File: lib/helpers.js [hunk 5]
```diff
@@ -72,3 +72,17 @@ exports.generateNonce = async ({entryStyle = 'human'} = {}) => {
   throw new Error(`Invalid "entryStyle" "${entryStyle}"; "entryStyle" must ` +
     `be "human" or "machine".`);
 };
+
+exports.combinedHash = ({clientId, hash} = {}) => {
+  if(clientId) {
+    return exports.prefixedHash(`${hash}:${clientId}`);
+  } else {
+    return exports.prefixedHash(hash);
+  }
```
Comment (dlongley) [inline lib/helpers.js:81]:
```suggestion
  }
  return exports.prefixedHash(hash);
```

File: lib/index.js [hunk 6]
```diff
@@ -550,8 +546,12 @@ api.verify = async ({
             });
         }
         tokenValue = token;
+        break;
       }
     }
+    if(!nonceVerified) {
+      return false;
```
Comment (dlongley) [inline lib/index.js:553]:
```suggestion
      // no token matches the given `hash` and `clientId`, so not verified
      return false;
```

File: lib/helpers.js [hunk 7]
```diff
@@ -72,3 +72,17 @@ exports.generateNonce = async ({entryStyle = 'human'} = {}) => {
...
+};
+
+exports.verifyHash = ({clientId, hash, sha256} = {}) => {
+  return crypto.timingSafeEqual(
+    new Buffer(sha256, 'base64'),
+    new Buffer(exports.combinedHash({clientId, hash}), 'base64'));
```
Comment (mattcollier) [inline lib/helpers.js:87]: I see this was copy/paste from existing code, however, I believe this syntax generates a deprecation warning like: 
```
(node:32) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
```
I think you'll want to use `Buffer.from`.

**Review by mattcollier — CHANGES_REQUESTED** [review]
Needs a changelog.

File: lib/index.js (hunk 2 above)
Comment (JoshDunnDev) [reply lib/index.js]: I changed the getAll function to return an object instead:
`
{
    allTokens,
    filteredTokens,
    expiredTokens
}
`

There is currently no use for `expiredTokens` at the moment, so should we return that data just incase we need it for future use?

Comment (dlongley) [reply lib/index.js]: Yes, let's just return it -- and perhaps call `filteredTokens` something like `unexpiredTokens`? (or a better name if you have one, but filtered is pretty non-descript).

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,17 @@
 # bedrock-authn-token ChangeLog
 
+## 4.0.0 - 2021-04-33
+### Added
+- **BREAKING**: Added `combinedHash` function to assist with concatenating `clientId` and `hash` for `prefixedHash`.
+- **BREAKING**: Added `verifyHash` functions to assist with comparing `prefixedHash`, and `sha256`.
```
Comment (dlongley) [inline CHANGELOG.md:6]: These aren't public APIs so they shouldn't be mentioned in the changelog -- it will be confusing.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,17 @@
...
+
+### Changed
+- **BREAKING**: Updated `getAll` function to now return an object of `allTokens`, `filteredTokens`, and `expiredTokens`.
+- **BREAKING**: Updated `set` api to use the updated object from the `getAll` api request for `tokens`.
+- Changed `filterExpiredTokens` to `true` in the `getAll` api request.
```
Comment (dlongley) [inline CHANGELOG.md:11]: Again, some of this has to do with implementation details -- the changelog should only mention bug fixes and changes to the public API.

File: lib/helpers.js [hunk 10]
```diff
@@ -72,3 +72,16 @@ exports.generateNonce = async ({entryStyle = 'human'} = {}) => {
...
+  if(clientId) {
+    return exports.prefixedHash(`${hash}:${clientId}`);
+  }
+  return exports.prefixedHash(hash);
+};
+
+exports.verifyHash = ({clientId, hash, sha256} = {}) => {
+  return crypto.timingSafeEqual(
+    new Buffer.from(sha256, 'base64'),
+    new Buffer.from(exports.combinedHash({clientId, hash}), 'base64'));
```
Comment (dlongley) [inline lib/helpers.js:86]:
```suggestion
    Buffer.from(sha256, 'base64'),
    Buffer.from(exports.combinedHash({clientId, hash}), 'base64'));
```

File: lib/index.js [hunk 11]
```diff
@@ -126,15 +127,15 @@ api.set = async ({
     try {
       // check if nonce exists
       tokens = await api.getAll({
```
Comment (dlongley) [inline lib/index.js:129]: Maybe this should be called `result` instead of `tokens` now.

File: lib/index.js [hunk 12]
```diff
@@ -125,9 +127,11 @@ api.set = async ({
     const {maxNonceCount} = cfg.nonce.defaults;
     try {
       // check if nonce exists
-      tokens = await api.getAll({
-        actor: null, account, email, type, filterExpiredTokens: false
+      const result = await api.getAll({
+        actor: null, account, email, type, filterExpiredTokens: true
       });
+      allTokens = result.allTokens;
+      tokens = result.tokens;
```
Comment (dlongley) [inline lib/index.js:134]:
```suggestion
      ({tokens, allTokens} = await api.getAll({
        actor: null, account, email, type, filterExpiredTokens: true
      }));
```

File: lib/index.js [hunk 13]
```diff
@@ -227,7 +224,8 @@ api.set = async ({
   if(type === 'nonce') {
     // include old value in the query to ensure that no concurrent
     // changes occurred via another process
-    const oldValue = tokens.length > 0 ? tokens.slice() : {$in: [null, []]};
+    const oldValue = allTokens.length > 0 ? allTokens
+      .slice() : {$in: [null, []]};
```
Comment (dlongley) [inline lib/index.js:228]: Split on conceptual parts of the statement if possible, it make it easier to read. That means, if your statement looks like this:

`<set this var equal to> <conditional> <this value if true> <this value if false>`

You'd break somewhere between the angled brackets, not within them -- unless there's no other choice.

```suggestion
    const oldValue = allTokens.length > 0 ?
      allTokens.slice() : {$in: [null, []]};
```

File: lib/index.js [hunk 14]
```diff
@@ -407,13 +405,28 @@ api.getAll = async ({
...
-      return token.expires > now;
-    });
+  for(const token of result) {
+    if(token.expires > now) {
+      if(filterExpiredTokens) {
+        tokens.push(token);
+      }
+    } else {
+      expiredTokens.push(token);
+    }
```
Comment (dlongley) [inline lib/index.js:419]: So looking at this more, I think continuing to have the `filterExpiredTokens` flag is likely to confuse devs.

If we're going to always run this loop that filters the output and then always return `allTokens`, `tokens`, and `expiredTokens` -- what's the point of the flag? Depending on how you set it the value of just `tokens` will change? I expect that to be confusing. Can we remove the `filterExpiredTokens` flag and adjust whatever else was calling this before to just use the appropriate value out of `result`?

For example, if something was setting `filterExpiredTokens: false` then it should just use `allTokens`, it should not make `allTokens` and `tokens` have the same value via a flag and then use `tokens`. If something was setting `filterExpiredTokens: true`, then it should just use `tokens` if all it needs are the unexpired ones.

**Review by dlongley — CHANGES_REQUESTED** [review]
Ok, nearly there. I think we should simplify the `getAll` function as noted in my comment and then this will be ready go to. Thanks!

File: lib/index.js (hunk 14 above)
Comment (JoshDunnDev) [reply lib/index.js:419]: Well we don't currently have anything setting `filterExpiredTokens` to `false` any longer. So if there is not a use case currently for that then I think removing the flag makes sense. However, if there is a use case for that, we are still going to have some sort of flag that specifies whether to use "filter" the tokens or not right? I think it's more confusing to move this out of the getAll function. I think getAll needs to be doing to filtering as that makes the most sense to me.

Comment (dlongley) [reply lib/index.js:417]: @JoshDunnDev, I'm not suggesting we move the filtering out, I'm saying that we don't need the flag. We should just make it so that the return value is *always*:

```js
{
  allTokens, // always all the tokens, not filtered
  tokens, // always all the unexpired tokens, does not change based on a flag
  expiredTokens, // always all the expired tokens
}
```

Comment (JoshDunnDev) [reply lib/index.js:412]: Alright I have gone ahead and made the change to remove the flag.

**Review by dlongley — APPROVED** [review]
Thanks!

**Review by mattcollier — APPROVED** [review]
TY
